src/Microsoft.AspNetCore.OData/EnableQueryAttribute.cs

```diff
@@ -38,6 +41,177 @@ public partial class EnableQueryAttribute : ActionFilterAttribute
         // also clearly ties the property to code in this assembly.
         private const string ModelKeyPrefix = "Microsoft.AspNet.OData.Model+";
 
+        /// <summary>
+        /// Marks if the query validation was run before the action execution. This is not always possible.
+        /// For cases where the run failed before action execution. We will run validation on result.
+        /// </summary>
+        private bool _queryValidationRunBeforeActionExecution;
+
+        /// <summary>
+        /// Stores the processed query options to be used later if OnActionExecuting was able to verify the query.
+        /// This is because ValidateQuery internally modifies query options (expands are prime example of this).
+        /// </summary>
+        private ODataQueryOptions _processedQueryOptions;
+
+        /// <summary>
+        /// Performs query validations before action is executed.
+        /// </summary>
+        /// <param name="context">Action context.</param>
+        public override void OnActionExecuting(ActionExecutingContext context)
+        {
+            if (context == null)
+            {
+                throw Error.ArgumentNull(nameof(context));
+            }
+
+            base.OnActionExecuting(context);
+
```

Comment:
The `_queryValidationRunBeforeActionExecution` field is set to `false` in quite a few places. Maybe we that can be set as the initial value here

Reply:
It is a boolean so it is by default set to false, but I am explicitly setting it for clarity sake. I can remove it if it is creating problems in reading.